[docs] Update Push notifications and expo-notifications docs (#20270)
expo/expo: docs/pages/push-notifications/sending-notifications.mdx

---
title: Send notifications with Expo's Push API
sidebar_title: Send notifications with Expo's Push API
---

import ImageSpotlight from '~/components/plugins/ImageSpotlight';
import { CODE } from '~/ui/components/Text';
import { BoxLink } from '~/ui/components/BoxLink';

The [`expo-notifications`](/versions/latest/sdk/notifications) library provides all the client-side functionality for push notifications. Expo also handles sending these notifications off to FCM and APNs. All you need to do is send the request to the Expo's Push API with the `ExpoPushToken` you grabbed in the last step.

> If you'd rather build a server that communicates with APNs and FCM directly, see [Send notifications with FCM and APNs](/push-notifications/sending-notifications-custom/). It's more complicated than using Expo's push notification service.

<ImageSpotlight
  alt="Diagram explaining sending a push from your server to device"
  src="/static/images/sending-notification.png"
  containerStyle={{ backgroundColor: '#fff' }}
/>

## Send push notifications using a server

> After you have setup your push notification credentials and added logic to get the `ExpoPushToken`, you can send it to the Expo API using an HTTPS POST request. You can do this by setting up a server with a database (or you can also write a command line tool to send them or send them straight from your app).

Reviewer: ```suggestion
After you setup your push notification credentials and add logic to get the `ExpoPushToken`, you can send it to the Expo API using an HTTPS POST request. You can do this by setting up a server with a database (or you can also write a command line tool to send them or send them straight from your app).
```

| [expo-server-sdk-rust](https://github.com/expo/expo-server-sdk-rust)                     | Rust     | Community     |
| [ExpoNotificationsBundle](https://github.com/solvecrew/ExpoNotificationsBundle)          | Symfony  | SolveCrew     |
| [exponent-server-sdk-php](https://github.com/Alymosul/exponent-server-sdk-php)           | PHP      | Community     |
| [expo-server-sdk-php](https://github.com/ctwillie/expo-server-sdk-php)                   | PHP      | Community     |
| [exponent-server-sdk-golang](https://github.com/oliveroneill/exponent-server-sdk-golang) | Golang   | Community     |
| [exponent-server-sdk-elixir](https://github.com/rdrop/exponent-server-sdk-elixir)        | Elixir   | Community     |
| [expo-server-sdk-dotnet](https://github.com/glyphard/expo-server-sdk-dotnet)             | dotnet   | Community     |
| [expo-server-sdk-java](https://github.com/jav/expo-server-sdk-java)                      | Java     | Community     |
| [laravel-expo-notifier](https://github.com/YieldStudio/laravel-expo-notifier)            | Laravel  | Community     |

> Each of the example servers above is a wrapper around the Expo's Push API.

Reviewer: ```suggestion
Each of the example servers above is a wrapper around Expo's Push API.
```

### Retry on failure

The first step of sending push notifications is to deliver them to the Expo push notification service, which internally adds them to a queue for delivery to Google (FCM), Apple (APNs), or other push notification providers. This first step can fail for several reasons:

- network issues between your server and the Expo push notification service
- an outage or degraded availability of the Expo notification service
- misconfigured push credentials
- an invalid notification payload

Some of these failures are temporary. For example, if the Expo push notification service is down or unreachable and you get a network error, an HTTP 429 error (Too Many Requests), or an HTTP 5xx error (Server Errors), use [exponential backoff](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/) to wait a few seconds before retrying. If the first retry attempt is unsuccessful, wait for longer (follow exponential backoff) and retry again. This lets the temporarily unavailable service recover before you retry.

> Other failures will not resolve themselves, no matter how many times you retry. For example, if your push notification payload is malformed, you may get an HTTP 400 response explaining the issue with the payload. You will also get an error if there are no push credentials for your project or if you send push notifications for different projects in the same request. You must fix the underlying causes of these issues to resolve them.

Reviewer: ```suggestion
Other failures will not resolve themselves. For example, if your push notification payload is malformed, you may get an HTTP 400 response explaining the issue with the payload. You will also get an error if there are no push credentials for your project or if you send push notifications for different projects in the same request.
```
